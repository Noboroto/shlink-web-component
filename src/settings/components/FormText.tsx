import type { FC, PropsWithChildren } from 'react';

export const FormText: FC<PropsWithChildren> = ({ children }) => (
  <small className="form-text text-muted d-block">{children}</small>
);
