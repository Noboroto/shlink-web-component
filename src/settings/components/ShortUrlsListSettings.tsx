import { LabeledFormGroup, OrderingDropdown, SimpleCard } from '@shlinkio/shlink-frontend-kit';
import type { FC } from 'react';
import type { ShortUrlsListSettings as ShortUrlsSettings } from '..';
import { useSetting } from '..';

export type ShortUrlsListSettingsProps = {
  updateShortUrlsListSettings: (settings: ShortUrlsSettings) => void;
  defaultOrdering: NonNullable<ShortUrlsSettings['defaultOrdering']>;
};

const SHORT_URLS_ORDERABLE_FIELDS = {
  dateCreated: 'Created at',
  shortCode: 'Short URL',
  longUrl: 'Long URL',
  title: 'Title',
  visits: 'Visits',
};

export const ShortUrlsListSettings: FC<ShortUrlsListSettingsProps> = (
  { updateShortUrlsListSettings, defaultOrdering },
) => {
  const shortUrlsList = useSetting('shortUrlsList');

  return (
    <SimpleCard title="Short URLs list" className="h-100">
      <LabeledFormGroup noMargin label="Default ordering for short URLs list:">
        <OrderingDropdown
          items={SHORT_URLS_ORDERABLE_FIELDS}
          order={shortUrlsList?.defaultOrdering ?? defaultOrdering}
          onChange={(field, dir) => updateShortUrlsListSettings({ defaultOrdering: { field, dir } })}
        />
      </LabeledFormGroup>
    </SimpleCard>
  );
};
