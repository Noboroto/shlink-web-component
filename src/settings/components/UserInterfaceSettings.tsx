import { faMoon, faSun } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import type { Theme } from '@shlinkio/shlink-frontend-kit';
import { getSystemPreferredTheme, SimpleCard, ToggleSwitch } from '@shlinkio/shlink-frontend-kit';
import type { FC } from 'react';
import { useMemo } from 'react';
import { useSetting } from '..';
import type { UiSettings } from '../types';

interface UserInterfaceProps {
  updateUiSettings: (settings: UiSettings) => void;

  /* Test seam */
  _matchMedia?: typeof window.matchMedia;
}

export const UserInterfaceSettings: FC<UserInterfaceProps> = ({ updateUiSettings, _matchMedia }) => {
  const ui = useSetting('ui');
  const currentTheme = useMemo(() => ui?.theme ?? getSystemPreferredTheme(_matchMedia), [ui?.theme, _matchMedia]);

  return (
    <SimpleCard title="User interface" className="h-100">
      <FontAwesomeIcon icon={currentTheme === 'dark' ? faMoon : faSun} className="float-end mt-1" />
      <ToggleSwitch
        checked={currentTheme === 'dark'}
        onChange={(useDarkTheme) => {
          const theme: Theme = useDarkTheme ? 'dark' : 'light';
          updateUiSettings({ ...ui, theme });
        }}
      >
        Use dark theme.
      </ToggleSwitch>
    </SimpleCard>
  );
};
