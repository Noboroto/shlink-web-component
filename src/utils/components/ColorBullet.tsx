import './ColorBullet.scss';

interface ColorBulletProps {
  color: string;
}

export const ColorBullet = ({ color }: ColorBulletProps) => (
  <div style={{ backgroundColor: color }} className="color-bullet" />
);
