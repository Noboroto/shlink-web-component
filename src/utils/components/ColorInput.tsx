import { useElementRef } from '@shlinkio/shlink-frontend-kit';
import type { FC } from 'react';
import { Input, InputGroup } from 'reactstrap';
import type { ColorPickerProps } from './ColorPicker';
import { ColorPicker } from './ColorPicker';

export const ColorInput: FC<Omit<ColorPickerProps, 'className'>> = ({ color, onChange }) => {
  const colorPickerRef = useElementRef<HTMLInputElement>();

  return (
    <InputGroup>
      <ColorPicker color={color} onChange={onChange} className="input-group-text" ref={colorPickerRef} />
      <Input value={color} readOnly onClick={() => colorPickerRef.current?.click()} />
    </InputGroup>
  );
};
