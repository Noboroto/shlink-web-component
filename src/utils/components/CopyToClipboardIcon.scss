.copy-to-clipboard-icon {
  cursor: pointer;
  font-size: 1.2rem;
}
