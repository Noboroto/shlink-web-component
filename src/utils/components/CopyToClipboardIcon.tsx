import { faClone as copyIcon } from '@fortawesome/free-regular-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import type { FC } from 'react';
import type { CopyToClipboardOptions } from '../helpers/clipboard';
import { copyToClipboard as defaultCopyToClipboard } from '../helpers/clipboard';
import { UnstyledButton } from './UnstyledButton';

type CopyToClipboardIconProps = CopyToClipboardOptions & {
  copyToClipboard?: typeof defaultCopyToClipboard;
};

export const CopyToClipboardIcon: FC<CopyToClipboardIconProps> = (
  { text, onCopy, copyToClipboard = defaultCopyToClipboard },
) => (
  <UnstyledButton
    className="ms-2 p-0"
    aria-label={`Copy ${text} to clipboard`}
    onClick={() => copyToClipboard({ text, onCopy })}
  >
    <FontAwesomeIcon icon={copyIcon} className="fs-5" />
  </UnstyledButton>
);
