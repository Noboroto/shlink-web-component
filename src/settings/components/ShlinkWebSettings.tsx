import { mergeDeepRight } from '@shlinkio/data-manipulation';
import { NavPillItem, NavPills } from '@shlinkio/shlink-frontend-kit';
import type { FC, ReactNode } from 'react';
import { useCallback } from 'react';
import { Navigate, Route, Routes } from 'react-router-dom';
import type { DeepPartial } from '../../utils/types';
import { SettingsProvider } from '..';
import type { RealTimeUpdatesSettings, Settings, ShortUrlsListSettings } from '../types';
import { RealTimeUpdatesSettings as RealTimeUpdates } from './RealTimeUpdatesSettings';
import { ShortUrlCreationSettings as ShortUrlCreation } from './ShortUrlCreationSettings';
import { ShortUrlsListSettings as ShortUrlsList } from './ShortUrlsListSettings';
import { TagsSettings as Tags } from './TagsSettings';
import { UserInterfaceSettings } from './UserInterfaceSettings';
import { VisitsSettings as Visits } from './VisitsSettings';

export type ShlinkWebSettingsProps = {
  settings: Settings;
  defaultShortUrlsListOrdering: NonNullable<ShortUrlsListSettings['defaultOrdering']>;
  updateSettings: (settings: Settings) => void;
};

const SettingsSections: FC<{ items: ReactNode[] }> = ({ items }) => (
  <>
    {items.map((child, index) => <div key={index} className="mb-3">{child}</div>)}
  </>
);

export const ShlinkWebSettings: FC<ShlinkWebSettingsProps> = (
  { settings, updateSettings, defaultShortUrlsListOrdering },
) => {
  const updatePartialSettings = useCallback(
    (partialSettings: DeepPartial<Settings>) => updateSettings(mergeDeepRight(settings, partialSettings)),
    [settings, updateSettings],
  );
  const toggleRealTimeUpdates = useCallback(
    (enabled: boolean) => updatePartialSettings({ realTimeUpdates: { enabled } }),
    [updatePartialSettings],
  );
  const setRealTimeUpdatesInterval = useCallback(
    (interval: number) => updatePartialSettings({ realTimeUpdates: { interval } as RealTimeUpdatesSettings }),
    [updatePartialSettings],
  );
  const updateSettingsProp = useCallback(
    <Prop extends keyof Settings>(prop: Prop, value: Settings[Prop]) => updatePartialSettings({ [prop]: value }),
    [updatePartialSettings],
  );

  return (
    <SettingsProvider value={settings}>
      <NavPills className="mb-3">
        <NavPillItem to="general">General</NavPillItem>
        <NavPillItem to="short-urls">Short URLs</NavPillItem>
        <NavPillItem to="other-items">Other items</NavPillItem>
      </NavPills>

      <Routes>
        <Route
          path="general"
          element={(
            <SettingsSections
              items={[
                <UserInterfaceSettings updateUiSettings={(v) => updateSettingsProp('ui', v)} />,
                <RealTimeUpdates
                  toggleRealTimeUpdates={toggleRealTimeUpdates}
                  setRealTimeUpdatesInterval={setRealTimeUpdatesInterval}
                />,
              ]}
            />
          )}
        />
        <Route
          path="short-urls"
          element={(
            <SettingsSections
              items={[
                <ShortUrlCreation updateShortUrlCreationSettings={(v) => updateSettingsProp('shortUrlCreation', v)} />,
                <ShortUrlsList
                  defaultOrdering={defaultShortUrlsListOrdering}
                  updateShortUrlsListSettings={(v) => updateSettingsProp('shortUrlsList', v)}
                />,
              ]}
            />
          )}
        />
        <Route
          path="other-items"
          element={(
            <SettingsSections
              items={[
                <Tags updateTagsSettings={(v) => updateSettingsProp('tags', v)} />,
                <Visits updateVisitsSettings={(v) => updateSettingsProp('visits', v)} />,
              ]}
            />
          )}
        />
        <Route path="*" element={<Navigate replace to="general" />} />
      </Routes>
    </SettingsProvider>
  );
};
