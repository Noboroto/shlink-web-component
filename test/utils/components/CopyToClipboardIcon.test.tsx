import { CopyToClipboardIcon } from '../../../src/utils/components/CopyToClipboardIcon';
import { checkAccessibility } from '../../__helpers__/accessibility';
import { renderWithEvents } from '../../__helpers__/setUpTest';

describe('<CopyToClipboardIcon />', () => {
  const copyToClipboard = vi.fn();
  const onCopy = vi.fn();
  const setUp = (text = 'foo') => renderWithEvents(
    <CopyToClipboardIcon text={text} onCopy={onCopy} copyToClipboard={copyToClipboard} />,
  );

  it('passes a11y checks', () => checkAccessibility(setUp()));

  it('wraps expected components', () => {
    const { container } = setUp();
    expect(container).toMatchSnapshot();
  });

  it.each([
    ['text'],
    ['bar'],
    ['baz'],
  ])('copies content to clipboard when clicked', async (text) => {
    const { user, container } = setUp(text);

    expect(copyToClipboard).not.toHaveBeenCalled();
    container.firstElementChild && await user.click(container.firstElementChild);
    expect(copyToClipboard).toHaveBeenCalledWith({ text, onCopy });
  });
});
