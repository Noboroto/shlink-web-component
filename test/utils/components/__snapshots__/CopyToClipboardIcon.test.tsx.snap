// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<CopyToClipboardIcon /> > wraps expected components 1`] = `
<div>
  <button
    aria-label="Copy foo to clipboard"
    class="border-0 ms-2 p-0"
    style="background-color: inherit; font-weight: inherit;"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-clone fs-5"
      data-icon="clone"
      data-prefix="far"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M64 464H288c8.8 0 16-7.2 16-16V384h48v64c0 35.3-28.7 64-64 64H64c-35.3 0-64-28.7-64-64V224c0-35.3 28.7-64 64-64h64v48H64c-8.8 0-16 7.2-16 16V448c0 8.8 7.2 16 16 16zM224 304H448c8.8 0 16-7.2 16-16V64c0-8.8-7.2-16-16-16H224c-8.8 0-16 7.2-16 16V288c0 8.8 7.2 16 16 16zm-64-16V64c0-35.3 28.7-64 64-64H448c35.3 0 64 28.7 64 64V288c0 35.3-28.7 64-64 64H224c-35.3 0-64-28.7-64-64z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;
